File: stripe/src/main/java/com/stripe/android/view/AuthActivityStarter.java
package com.stripe.android.view;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import java.lang.ref.WeakReference;

public interface AuthActivityStarter<StartDataType> {
    void start(@NonNull StartDataType data);

    /**
     * A representation of an object (i.e. Activity or Fragment) that can start an activity.
     */
    final class Host {
        @NonNull private final WeakReference<Activity> mActivityRef;
        @Nullable private final WeakReference<Fragment> mFragmentRef;

        @NonNull
        public static Host create(@NonNull Fragment fragment) {
            return new Host(fragment.requireActivity(), fragment);
        }

        public static Host create(@NonNull Activity activity) {
            return new Host(activity, null);
        }

        private Host(@NonNull Activity activity, @Nullable Fragment fragment) {
            this.mActivityRef = new WeakReference<>(activity);
            this.mFragmentRef = fragment != null ? new WeakReference<>(fragment) : null;
        }

        @Nullable
        public Activity getActivity() {
            return mActivityRef.get();
        }

        public void startActivityForResult(@NonNull Class target,
                                           @NonNull Bundle extras,
                                           int requestCode) {
            final Activity activity = mActivityRef.get();
            if (activity == null) {
                return;
            }

            final Intent intent = new Intent(activity, target).putExtras(extras);

            if (mFragmentRef != null) {
                final Fragment fragment = mFragmentRef.get();
                if (fragment != null) {
                    fragment.startActivityForResult(intent, requestCode);
                }
            } else {
                activity.startActivityForResult(intent, requestCode);
            }
        }
    }
}

File: stripe/src/main/java/com/stripe/android/view/AuthActivityStarter.kt
package com.stripe.android.view

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import java.lang.ref.WeakReference

internal interface AuthActivityStarter<ArgsType> {
    fun start(args: ArgsType)

    /**
     * A representation of an object (i.e. Activity or Fragment) that can start an activity.
     */
    class Host internal constructor(
        activity: Activity,
        fragment: Fragment?,
        val statusBarColor: Int?
    ) {
        private val activityRef: WeakReference<Activity> = WeakReference(activity)
        private val fragmentRef: WeakReference<Fragment>? = fragment?.let { WeakReference(it) }

        internal val activity: Activity?
            get() = activityRef.get()

        internal val fragment: Fragment?
            get() = fragmentRef?.get()

        internal fun startActivityForResult(
            target: Class<*>,
            extras: Bundle,
            requestCode: Int
        ) {
            val activity = activityRef.get() ?: return

            val intent = Intent(activity, target).putExtras(extras)

            if (fragmentRef != null) {
                val fragment = fragmentRef.get()
                if (fragment?.isAdded == true) {
                    fragment.startActivityForResult(intent, requestCode)
                }
            } else {
                activity.startActivityForResult(intent, requestCode)
            }
        }

        internal companion object {
            @JvmSynthetic
            internal fun create(
                fragment: Fragment
            ): Host {
                val activity = fragment.requireActivity()
                return Host(
                    activity = activity,
                    fragment = fragment,
                    statusBarColor = activity.window?.statusBarColor
                )
            }

            @JvmSynthetic
            internal fun create(
                activity: Activity
            ): Host {
                return Host(
                    activity = activity,
                    fragment = null,
                    statusBarColor = activity.window?.statusBarColor
                )
            }
        }
    }
}
